### PR TITLE
Add gNMI History extension documentation

### PR DESCRIPTION
**Pull request by jcostaroberts**
This change adds the specification for the gNMI History extension. This extension lets clients request historical data.

### CODE REVIEW

**Comment by jcostaroberts** [comment]
@googlebot I signed it!

**Comment by jcostaroberts** [comment]
@googlebot I signed it!

File: rpc/gnmi/gnmi-history.md
Comment (robshakir) [inline rpc/gnmi/gnmi-history.md:1]: (nit): Can you 80ch wrap the doc by any chance? It just makes things easier to review in the future.

File: rpc/gnmi/gnmi-history.md
Comment (robshakir) [inline rpc/gnmi/gnmi-history.md:41]: I wonder whether it's a good idea to wait until the specified time has been reached -- it seems like a potential DoS vector if a client can send a lot of requests that potentially are held open for a long time. Is there a specific use case that motivates this?

File: rpc/gnmi/gnmi-history.md
Comment (robshakir) [inline rpc/gnmi/gnmi-history.md:47]: Similar question here as to whether a start time in the future should be supported.

**Review by robshakir — COMMENTED** [review]
I have one clarification question here -- but this LGTM.

@mkhsueh @gscl -- PTAL and see whether you're good with this.

File: rpc/gnmi/gnmi-history.md
Comment (mkhsueh) [inline rpc/gnmi/gnmi-history.md:45]: (nit): Consider calling out the "between" behavior as the half-open interval, perhaps at the start of this paragraph; the spec mentions it thrice before defining semantics.

File: rpc/gnmi/gnmi-history.md
Comment (jcostaroberts) [reply rpc/gnmi/gnmi-history.md:41]: We went back and forth on this. The reason to wait is that it seems undesirable for a client request ending at t=time.Now() to result in an error just because the server's time is slightly behind the client's. Another approach we considered is to have the server immediately return all updates between the specified start time and its own current time u (u < t). This also seems undesirable because the client might well use time t as the start time for a later request (if, say, it does a get every hour), which would create a window between u and t for which it would not receive updates, and wouldn't know it hadn't.

The approach we landed on was either to give the client exactly what it asked for or to return an error, and to leave it up to the server whether it wants to wait. In practice I imagine a server would have some short period it would be willing to wait, and otherwise would return an error.

What do you think?

File: rpc/gnmi/gnmi-history.md
Comment (jcostaroberts) [reply rpc/gnmi/gnmi-history.md:1]: Done.

File: rpc/gnmi/gnmi-history.md
Comment (jcostaroberts) [reply rpc/gnmi/gnmi-history.md]: Done.

File: rpc/gnmi/gnmi-history.md
Comment (robshakir) [reply rpc/gnmi/gnmi-history.md]: That sounds reasonable to me -- why don't we add a small note here that just explains that a client should expect that there are conditions in which:

* Connections with start times that that are far into the future may be rejected.
* Servers may limit the number of connections of this nature.

And notes that server implementors should be careful not to cause themselves a resource exhaustion problem here. It's stepping beyond a spec into implementation advice, but I think we did something similar in the 'Considerations for using Get' section of the main spec.

Comment (jcostaroberts) [reply rpc/gnmi/gnmi-history.md]: Yeah that seems good. I've added a section at the end of the document including these points.

File: rpc/gnmi/gnmi-history.md
Comment (mhines01) [inline rpc/gnmi/gnmi-history.md:58]: Does the stream include an initial snapshot? or only updates received between the date ranges?

File: rpc/gnmi/gnmi-history.md
Comment (mhines01) [inline rpc/gnmi/gnmi-history.md:31]:
```proto
message History {
  oneof request {
    int64 timestamp = 1;
    TimeRange range = 2;
  }
}
```

might clearer on what the timestamp represents - when reading this the first time i thought it was a start without end vs. a stream or snapshot

File: rpc/gnmi/gnmi-history.md
Comment (mhines01) [inline rpc/gnmi/gnmi-history.md:71]: I feel like this going to be a little more complicated in the expectations from the service providing the stream...

If you have something that both going to serve historical data as well as current data how the data is muxed into those streams is likely to be complicated.

If you want it easy on the client - you would say start at the beginning and just play our the stream from history and then once done start sending all future updates from the live stream but that could have huge buffering problems on the server side of things.

The other option is to allow the streams to be mixed and just play the updates out of order but that would like cause clients problems as they wouldn't know when the "history" stream is actually done or they would just drop all non latest updates as currently for the clients i use - receiving out of order updates -- the old ones are just dropped since they have been replaced by new data anyway

File: rpc/gnmi/gnmi-history.md
Comment (jcostaroberts) [reply rpc/gnmi/gnmi-history.md:58]: I tried to address this below, with "Whether the server should transmit the specified data tree as it existed at the start time is determined as usual by the `updates_only` field on the `SubscriptionList` message."

Maybe a clearer way to say this is "As usual, the server should also transmit the specified data tree as it existed at the subscription start time unless the `updates_only` field on the `SubscriptionList` message is set." What do you think?

File: rpc/gnmi/gnmi-history.md
Comment (jcostaroberts) [reply rpc/gnmi/gnmi-history.md:71]: My reading of the gNMI spec is that updates needn't be buffered until the sync_response message has been sent:

"When the target has transmitted the initial updates for all paths specified within the subscription, a SubscribeResponse message with the sync_response field set to true MUST be transmitted to the client to indicate that the initial transmission of updates has concluded. This provides an indication to the client that all of the existing data for the subscription has been sent at least once. For STREAM subscriptions, such messages are not required for subsequent updates."

So that should cover this case as well.

Comment (robshakir) [reply rpc/gnmi/gnmi-history.md:71]: I think that in the scenario  that you have a client that wants to take both "historical and future state", then the behaviour that seems logical (and the lowest implementation overhead) is:

 * play all updates starting at `start_time`
 * once this has completed, send current version of the tree

For a history-consuming client, there is already a need to parse more than a single update to a path within the stream. For example, if we are drawing some graph, then we insert a point at all received datapoints. The buffering would ensure no discontinuities in this graph - but I suspect such a client should rather implement:

- history query for $start_time -> $now
- live subscription

if it is interested in both live and historical data. Or simply a history query if it already knows the time bounds.

Comment (jcostaroberts) [reply rpc/gnmi/gnmi-history.md:71]: I'm not sure I understand the first case you're describing--there's no combination of start and end time that should produce updates beginning at a historical start time and then the whole current tree. If updates_only isn't set on a TimeRange subscription, the server should send the tree as it existed at the specified start time, and then incremental updates from then up until the specified end.

In the second case, I agree that clients may just do a historical query + a normal subscription, but are you saying that something about the spec should change, or just pointing out that clients always have the option of implementing open-ended historical queries this way?

Comment (robshakir) [reply rpc/gnmi/gnmi-history.md:71]: I'm not sure that the "history + whole current tree" is forbidden in the current gNMI specification, right? I agree that the optimal approach is not to re-send the current tree (since this might be sending duplicates), but there's no specific requirement that a server MUST NOT ever send the same update twice (in-fact, this is required for a solely SAMPLE subscription). I was just pointing out that Marcus' concern on implementation complexity might be mitigated in this manner.

For the latter, I was also intending to support this implementation -- and was just trying to highlight that I think client implementation doesn't necessarily need to change here (for a use case where having not seen a particular set of updates that were dropped/suppressed by the server whilst replaying the history would have a material impact).

Comment (jcostaroberts) [reply rpc/gnmi/gnmi-history.md:71]: How would you specify "history + whole current tree"? The only way to get the full tree with a `STREAM` subscription is to set `updates_only` to `false`. If the History extension is used, this will return the tree at the `TimeRange`'s specified `start` time, and then any updates that occurred between that time and the specified `end` time. Thus the only way to get the current tree with the History extension is to specify a `start` time of time.Now, which means no historical data. So history + whole current tree is not possible.

My understanding of Marcus' concern is that it's about whether, in servicing a `STREAM` subscription with `start` < `now` < `end`, the server should buffer updates that occur between `now` and `end` until the initial state has been synchronized (i.e., until the server has sent a `sync_response`), or just transmit them as soon as they occur. That is, I think Marcus' point is about update ordering, not duplication. And by my reading, the spec doesn't prohibit sending current updates before the initial tree has synchronized.

File: rpc/gnmi/gnmi-history.md
Comment (robshakir) [reply rpc/gnmi/gnmi-history.md:58]: This sounds good to me.

File: rpc/gnmi/gnmi-history.md
Comment (robshakir) [reply rpc/gnmi/gnmi-history.md:71]: Apologies, I think we're talking past each other here. I am not saying that we need a way to be able to have the updates plus the whole current tree, I'm saying that if an implementation honoured a history approach by implementing this way then this would not be in violation of the spec.

Regarding the buffering, I agree with you - there is no requirement for ordering in the stream in the current gNMI. I don't see that we should introduce this now.

File: rpc/gnmi/gnmi-history.md
Comment (robshakir) [reply rpc/gnmi/gnmi-history.md:31]: Maybe even `at_timestamp` or `snapshot_timestamp` would be clearer. I agree with you that coming back this doesn't look as clear as it could do.

**Comment by robshakir** [comment]
@jcostaroberts - apologies for the delay here, this got tail-dropped with some personal circumstances that I had. This generally LGTM, if we can make the couple of clarifications above then I think we can merge it.

File: rpc/gnmi/gnmi-history.md
Comment (jcostaroberts) [reply rpc/gnmi/gnmi-history.md]: I've changed the field name here (and in gnmi_ext.proto) to `snapshot_time`. I like `at_time` better as a standalone name but making one field a prepositional phrase and the other a noun kind of stresses me out. And went with `time` over `timestamp` because to me "timestamp" indicates a time at which something _happened_, whereas the client here isn't indicating anything happening at all--it just wants the tree at the specified time. Let me know if you have any objection.

File: rpc/gnmi/gnmi-history.md
Comment (aashaikh) [inline rpc/gnmi/gnmi-history.md:51]: maybe clearer to say "... mode for a snapshot request must be ..."

File: rpc/gnmi/gnmi-history.md
Comment (aashaikh) [inline rpc/gnmi/gnmi-history.md:58]: I'd suggest to call this section "Range requests" or similar ; I think STREAM is just the type of subscription to use for timerange requests.

File: rpc/gnmi/gnmi-history.md
Comment (aashaikh) [inline rpc/gnmi/gnmi-history.md:67]: nit: suggest to omit "and then return" or maybe just say "and then close the RPC"

**Review by aashaikh — COMMENTED** [review]
This LGTM -- just added a couple of minor suggestions.  Leaving for  @robshakir to set approval bit.

File: rpc/gnmi/gnmi-history.md
Comment (jcostaroberts) [reply rpc/gnmi/gnmi-history.md]: Thanks--all good catches.

**Review by robshakir — APPROVED** [review]
LGTM, thanks for all the changes here.

**Comment by jcostaroberts** [comment]
Thanks! Are we also in a good place on https://github.com/openconfig/gnmi/pull/80 ?

**Comment by robshakir** [comment]
Yes, I merged this internally earlier and will do a push to the gNMI repo
this afternoon.

Thanks,
r.

**Pull request merged**
